In [29]:
# Instalación de dependencias (necesario para Colab)
try:
    from itertools import combinations
    from itertools import product
    import numpy as np
    import matplotlib.pyplot as plt
    print("Librerías ya instaladas.")
except ImportError:
    !pip install numpy matplotlib
    import numpy as np
    import matplotlib.pyplot as plt


Librerías ya instaladas.


Parte 1

In [30]:
def calcular_X(resultados):
    """
    resultados: lista o array de longitud 10 con valores 0 o 1
    """
    X = 0
    for i, Mi in enumerate(resultados, start=1):
        X += Mi / (2 ** i)
    return X

In [31]:
resultados = [1, 0, 1, 1, 0, 0, 1, 0, 1, 1]
X = calcular_X(resultados)
print(X)

0.6982421875


In [32]:
# Genera todas las combinaciones posibles de 0 y 1 para 3 lanzamientos
t = product([0, 1], repeat=3) 
print(list(t)) 

[(0, 0, 0), (0, 0, 1), (0, 1, 0), (0, 1, 1), (1, 0, 0), (1, 0, 1), (1, 1, 0), (1, 1, 1)]


In [33]:
# Genera todas las combinaciones posibles de "cara" y "número" para 10 lanzamientos y calculo x
recorrido = set()
for resultados in product([0, 1], repeat=10):
    X = calcular_X(resultados)
    recorrido.add(X)
elementos = len(recorrido)
print(f"Elementos distintos en el recorrido: {elementos}")        

Elementos distintos en el recorrido: 1024


In [34]:

def analizar_X(i):
    recorrido_x = set()
    for resultados in product([0, 1], repeat=i):
        X = calcular_X(resultados)
        recorrido_x.add(X)
    recorrido_x = list(recorrido_x)
    media = np.mean(recorrido_x)
    varianza = np.var(recorrido_x)
    return media, varianza


In [35]:
cantidades = [10, 12, 14, 16, 18, 20]

print(f"{'Monedas':<10} | {'Esperanza':<12} | {'Varianza':<12}")
print("-" * 40)

for k in cantidades:
    E, Var = analizar_X(k)
    print(f"{k:<10} | {E:<12.5f} | {Var:<12.5f}")

Monedas    | Esperanza    | Varianza    
----------------------------------------
10         | 0.49951      | 0.08333     
12         | 0.49988      | 0.08333     
14         | 0.49997      | 0.08333     
16         | 0.49999      | 0.08333     
18         | 0.50000      | 0.08333     
20         | 0.50000      | 0.08333     


Parte 2

In [48]:
def simular_X(monedas):
    resultados = np.random.randint(0, 2, size=monedas)
    return resultados


In [50]:
n = [10**3, 10**4, 10**5, 10**6]
simulaciones = []
def N_simulaciones(veces,monedas):
    simulacion = []
    for i in range(0,veces):
        simulacion.append(calcular_X(simular_X(monedas)))   
    return simulacion
for k in n:
    simulaciones.append(N_simulaciones(k,10))




In [42]:
def frecuencia_Relativa(lista):
    x=0
    for i in lista:
        if i >= 2/7 and i <= 6/7:
            x += 1
    return x / len(lista)

In [51]:
print(f"{'Simulacion':<10} | {'Frecuencia relativa':<12}")
print("-" * 40)

for k in range(0,4):
    F = frecuencia_Relativa(simulaciones[k])
    print(f"{n[k]:<10} | {F:<12.5} ")

Simulacion | Frecuencia relativa
----------------------------------------
1000       | 0.574        
10000      | 0.5643       
100000     | 0.57258      
1000000    | 0.57165      


In [ ]:
cantidades = [10, 12, 14, 16, 18, 20]
n = [10**3, 10**4, 10**5, 10**6]
import time

inicio = time.perf_counter()
print(inicio)
frecuencia_Relativa_10 = frecuencia_Relativa(N_simulaciones(n[0], cantidades[0]))
frecuencia_Relativa_12 = frecuencia_Relativa(N_simulaciones(n[1], cantidades[1]))
frecuencia_Relativa_14 = frecuencia_Relativa(N_simulaciones(n[2], cantidades[2]))
inicio2 = time.perf_counter()
print(inicio2 - inicio)
frecuencia_Relativa_16 = frecuencia_Relativa(N_simulaciones(n[3], cantidades[3]))
inicio3 = time.perf_counter()
print(inicio3 - inicio2)
frecuencia_Relativa_18 = frecuencia_Relativa(N_simulaciones(n[3], cantidades[4]))
inicio4 = time.perf_counter()
print(inicio4 - inicio3)
frecuencia_Relativa_20 = frecuencia_Relativa(N_simulaciones(n[3], cantidades[5]))
inicio5 = time.perf_counter()
print(inicio5 - inicio4)
resultados_frecuencia = [frecuencia_Relativa_10, frecuencia_Relativa_12, frecuencia_Relativa_14, frecuencia_Relativa_16, frecuencia_Relativa_18, frecuencia_Relativa_20]

print(f"{'monedas':<10} | {'Simulacion':<10} | {'Frecuencia relativa':<12}")
print("-" * 40)

for i in range(0,6):
    F = resultados_frecuencia[i]
    print(f"{cantidades[i]:<10} | {n[0]:<10} | {F:<12.5} ")

416860.6349718
1.217345299955923
11.141601400042418
11.941844799963292
12.111915400018916
monedas    | Simulacion | Frecuencia relativa
----------------------------------------
10         | 1000       | 0.557        
12         | 1000       | 0.5734       
14         | 1000       | 0.56939      
16         | 1000       | 0.57086      
18         | 1000       | 0.5717       
20         | 1000       | 0.57055      


8) Para comprobar que la esperanza y la varianza se aproximan hacia la uniforme continua realizamos 10000 simulaciones para cada cantidad de monedas. Así "quitamos la suerte" del experimento y estabilizamos los valores

In [ ]:
def analizar_monedas(k):

    n_simulaciones = 10000
    
    lanzamientos = np.random.randint(0, 2, size=(n_simulaciones, k))
    
    potenciasMi = 2 ** np.arange(1, k + 1)
    
    # Calculamos X para cada una de las 10,000 simulaciones
    # Dividimos cada moneda por su potencia y sumamos la fila
    simulaciones_X = np.sum(lanzamientos / potenciasMi, axis=1)
    
    # Esperanza y varianza de esos resultados
    media = np.mean(simulaciones_X)
    varianza = np.var(simulaciones_X)
    
    return media, varianza

In [ ]:

cantidades = [10, 20, 30, 40, 50]

print(f"{'Monedas':<10} | {'Esperanza':<12} | {'Varianza':<12}")
print("-" * 40)

for k in cantidades:
    E, Var = analizar_monedas(k)
    print(f"{k:<10} | {E:<12.5f} | {Var:<12.5f}")


Monedas    | Esperanza    | Varianza    
----------------------------------------
10         | 0.49505      | 0.08333     
20         | 0.49942      | 0.08241     
30         | 0.50349      | 0.08291     
40         | 0.49863      | 0.08315     
50         | 0.50124      | 0.08298     


Cómo se observa en la tabla, a medida que se aumenta la cantidad de monedas, los valores de Esperanza y Varianza se aproximan a 0.5 y 0.0833